core/src/main/scala/kafka/admin/AdminClient.scala

```diff
@@ -137,6 +151,71 @@ class AdminClient(val time: Time,
       broker -> Try[NodeApiVersions](new NodeApiVersions(getApiVersions(broker).asJava))
     }.toMap
 
+
+  /*
+   * Remove all the messages whose offset is smaller than the given offset of the corresponding partition
+   *
+   * PurgeDataResult contains either lowWatermark of the partition or exception. We list the possible exception
+   * and their interpretations below:
+   *
+   * - DisconnectException if leader node of the partition is not available. Need retry by user.
+   * - TopicAuthorizationException if the topic doesn't exist and the user doesn't have the authority to create the topic
+   * - TimeoutException if response is not available within the timeout specified by either Future's timeout or AdminClient's request timeout
+   * - UnknownTopicOrPartitionException if the partition doesn't exist or if the user doesn't have the authority to describe the topic
+   * - NotLeaderForPartitionException if broker is not leader of the partition. Need retry by user.
+   * - OffsetOutOfRangeException if the offset is larger than high watermark of this partition
+   *
+   */
+
+  def purgeDataBefore(purgeOffsets: Map[TopicPartition, Long]): Future[Map[TopicPartition, PurgeDataResult]] = {
```

Comment:
The vast majority of Kafka code is not public API. The somewhat unintuitive way that public API is defined at the moment is by what has javadoc published:

https://kafka.apache.org/0102/javadoc/index.html

Apart from that, the old producers and consumers are also public API (until we remove them). The admin package is definitely internal and that's why there was no KIP for the Scala AdminClient when it was added in 0.9.0.0 (or any of the changes since).